### DISCUSSION TITLE
Comparison of `Float` values

### DISCUSSION THREAD

**Discussion by C8620** [mitsuba-renderer/mitsuba3#1056] [Q&A] [answered]
I am in the middle of the process of implementing Irawan paper using Python custom BSDF according to the C++ code in OG Mitsuba, and during (painful) debugging I have noticed that some statements involving comparison is not working, one quick example:

```
import drjit as dr
import mitsuba as mi

mi.set_variant('cuda_ad_rgb') # Using CUDA_AD due to LLVM issues and scalar incompatability.

var1 = mi.Float(1.0)

if(var1 > 0.5):
    print("var1 > 0.5")
else:
    print("var1 <= 0.5")
```
would yield
```
Exception                                 Traceback (most recent call last)
Cell In[1], [line 8](vscode-notebook-cell:?execution_count=1&line=8)
      [4](vscode-notebook-cell:?execution_count=1&line=4) mi.set_variant('cuda_ad_rgb')
      [6](vscode-notebook-cell:?execution_count=1&line=6) var1 = mi.Float(1.0)
----> [8](vscode-notebook-cell:?execution_count=1&line=8) if(dr.sum(var1) > 0.5):
      [9](vscode-notebook-cell:?execution_count=1&line=9)     print("var1 > 0.5")
     [10](vscode-notebook-cell:?execution_count=1&line=10) else:

[446](https://file+.vscode-resource.vscode-cdn.net/b%3A/RRF/~/AppData/Local/Packages/PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0/LocalCache/local-packages/Python310/site-packages/drjit/router.py:446) def op_bool(self):
--> [447](https://file+.vscode-resource.vscode-cdn.net/b%3A/RRF/~/AppData/Local/Packages/PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0/LocalCache/local-packages/Python310/site-packages/drjit/router.py:447)     raise Exception(
    [448](https://file+.vscode-resource.vscode-cdn.net/b%3A/RRF/~/AppData/Local/Packages/PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0/LocalCache/local-packages/Python310/site-packages/drjit/router.py:448)         "To convert an Dr.Jit array into a boolean value, use a mask reduction "
    [449](https://file+.vscode-resource.vscode-cdn.net/b%3A/RRF/~/AppData/Local/Packages/PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0/LocalCache/local-packages/Python310/site-packages/drjit/router.py:449)         "operation such as drjit.all(), drjit.any(), drjit.none(). Special "
    [450](https://file+.vscode-resource.vscode-cdn.net/b%3A/RRF/~/AppData/Local/Packages/PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0/LocalCache/local-packages/Python310/site-packages/drjit/router.py:450)         "variants (drjit.all_nested(), etc.) are available for nested arrays.")

Exception: To convert an Dr.Jit array into a boolean value, use a mask reduction operation such as drjit.all(), drjit.any(), drjit.none(). Special variants (drjit.all_nested(), etc.) are available for nested arrays.
```

I understand I might have missed something, but in [Mask operations](https://drjit.readthedocs.io/en/latest/reference.html#mask-operations) there is no method to perform inequalities to re-implement some of the codes in Irawan. Is there any suitable way to go around this? Preferably, a way to transform `Float` in its array form to python `float`?

**Comment by rtabbara** [comment, marked as answer]
Hi @c86-moe,

You can use the `dr.select` mask operation when working with JitArray types

```python
# Arrays of width 2
a = mi.Float(1.0, 3.0)
b = mi.Float(2.0, 2.0)
c = mi.Float(3.0, 1.0)

# Will create a new array which when evaluated returns [3.0, 2.0]
d = dr.select(a > 1.0, b, c)
```

**Reply by C8620** [reply]
Hi @rtabbara 

Many thanks! Although this is not the direct answer, but I am able to write a little helper function to mitigate this:

```
def ineq(condition):
    return dr.any(dr.select(condition, True, False))
```

I believe this may not be the most elegent way, but for now it works.

Cheers!

**Reply by C8620** [reply]
Hi @rtabbara 

I was embedding such code into the longer piece and it appears that such actions may crash the jupyter notebook kernel:

```
def sample(self, ctx, si, sample1, sample2, active):
        print("🏝️ sampling 🏝️")
        cos_theta_i = mi.Frame3f.cos_theta(si.wi)
        bs = dr.zeros(mi.BSDFSample3f)
        if(dr.all(active)):
            bs.wo = warp.square_to_cosine_hemisphere(sample2)
            pdf = warp.square_to_cosine_hemisphere_pdf(bs.wo)
            bs.pdf = pdf
            bs.eta = 1.0
            bs.sampled_type = mi.UInt32(+mi.BSDFFlags.DiffuseReflection)
            bs.sampled_component = 0
            value = self.eval(ctx, si, bs.wo, active)
            if(self.ineq(pdf > 0.0)):
                return bs, mi.depolarizer(value)
        return bs, mi.Color3f(0.0)

        def ineq(self, condition):
            return dr.any(dr.select(condition, True, False))
```

The above code is an attempt of reimplementation in Python of the `/src/bsdfs/diffuse.cpp`'s sample function, and when I run this code, I will get prompt of a crashed kernel, and log says following:

```
15:42:30.284 [info] Handle Execution of Cells 2 for b:\RRF\irawan3.ipynb
15:42:30.287 [info] Kernel acknowledged execution of cell 2 @ 1706888550286
15:42:30.626 [error] Disposing session as kernel process died ExitCode: 3221226505, Reason: jit_llvm_init(): detected LLVM version lacks critical functionality, shutting down LLVM backend..
```

Any idea what might be the cause?

**Reply by rtabbara** [reply]
Hi @c86-moe ,

That sounds like an issue with your LLVM installation/setup. There's a similar issue reported [here](https://github.com/NVlabs/sionna/discussions/160) with a proposed solution that might fix your problem.

**Reply by C8620** [reply]
Hi @rtabbara,

Many thanks for your timely reply and helpful responses.

I am using CUDA backend and there is an nVidia graphics card installed (with CUDA toolkit also), should this mean LLVM is not needed? Above errors are all produced with `mi.set_variant('cuda_ad_rgb')`.

I also have manually set the path of 'LLVM-C.dll' (using windows) in both Windows system properties and using attempted `os.environ['DRJIT_LIBLLVM_PATH'] = 'C:\\Program Files\\LLVM\\bin\\LLVM-C.dll'` in previous or the same cell in jupyter notebook, but mitsuba3 seems still have troubles using it when setting variant to `llvm_ad_rgb`, reporting:

```
AttributeError: jit_init_thread_state(): the LLVM backend is inactive because the LLVM shared library ("LLVM-C.dll") could not be found! Set the DRJIT_LIBLLVM_PATH environment variable to specify its path.
```

Any help would be greatly appreciated.

Edit: I furthered on to use LLVM 16.0.4 instead and now the kernel would die with a different error:
`10:01:01.230 [error] Disposing session as kernel process died ExitCode: 3221226505, Reason: `

**Reply by rtabbara** [reply]
With the LLVM downgrade, it sounds like now you have an issue with the code itself and you'll have to bisect to find the relevant line causing the issue.
